stablehlo/reference/ProcessGrid.h

```diff
@@ -88,13 +89,23 @@ class ProcessGrid {
   ProcessGrid(uint32_t numReplicas, uint32_t numPartitions);
   /// @}
 
-  /// StableHLO `cross_replica` communication strategy.
-  ProcessGroups crossReplica(SmallVector<SmallVector<uint32_t>> replicaGroups);
-
   /// StableHLO `cross_partition` communication strategy.
   ProcessGroups crossPartition(
       SmallVector<SmallVector<uint32_t>> partitionGroups);
 
+  /// StableHLO `cross_replica` communication strategy.
+  ProcessGroups crossReplica(SmallVector<SmallVector<uint32_t>> replicaGroups);
+
+  /// Inserts `inputs` to StableHLO `outfeed`.
+  void outfeed(ArrayRef<Tensor> inputs);
+
+  /// Returns whether the outfeed queue is empty.
+  bool outfeedEmpty();
+
+  /// Pops the first entry in the outfeed queue and returns its contents. The
+  /// method fails with fatal error if the queue is empty.
+  SmallVector<Tensor> popOutfeed();
```

Comment:
Can we perhaps expose `getOutfeed` directly, so that reading it requires only 1 API call and doesn't involve mutating the underlying queue?

Reply:
Actually I realized that this operation is not thread safe so we should probably keep it the way it is and introduce a separate lock for it.

Reply:
That's a good catch! As far `getOutfeed` goes, it could take a lock, copy the queue into a separate SmallVector and return the copy. But this is moot at this point because testing will be happening in a follow-up, so this can be figured out then.